src/main/java/com/google/devtools/build/lib/actions/ActionCacheChecker.java

```diff
@@ -409,13 +487,21 @@ public void updateActionCache(
         actionCache.remove(execPath);
       }
       if (!metadataHandler.artifactOmitted(output)) {
-        // Output files *must* exist and be accessible after successful action execution. We use the
-        // 'constant' metadata for the volatile workspace status output. The volatile output
-        // contains information such as timestamps, and even when --stamp is enabled, we don't want
-        // to rebuild everything if only that file changes.
-        FileArtifactValue metadata = getMetadataOrConstant(metadataHandler, output);
-        Preconditions.checkState(metadata != null);
-        entry.addFile(output.getExecPath(), metadata, /* saveExecPath= */ false);
+        if (output.isTreeArtifact()) {
```

Comment:
We used to just remove the entry from cache on a miss -- why not keep it the same way?

Reply:
I didn't change this logic. Not sure How the line you commented on are related to this.